File: packages/sdk/src/canonical.ts
// Sorted-key canonical JSON serialization.
//
// This MUST produce byte-identical output to provider/src/canonical.rs.
// The provider signs `to_canonical_bytes(record)` with its Secure Enclave;
// the AppView verifies that signature by re-canonicalizing here and
// recomputing. Any divergence breaks every receipt.
//
// Rules (matched in canonical.rs):
//   * UTF-8 output, no BOM.
//   * Object keys sorted lexicographically (UTF-16 code unit order, which
//     matches Rust's String <-> String comparison for the BMP and is
//     identical for ASCII keys — every key in our lexicons is ASCII).
//   * No insignificant whitespace.
//   * Strings escaped with the smallest legal form: \", \\, \b, \f, \n,
//     \r, \t, and \u00XX (lowercase hex) for control characters.
//   * Numbers: integers only. Floats throw. Lexicon-defined numeric
//     fields are integer; we never sign over a float.
//   * Booleans: true / false.
//   * null is allowed.
//   * Bytes (signatures etc.) are base64-encoded by callers BEFORE
//     entering this serializer; this layer treats them as strings.

export class CanonicalError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "CanonicalError";
  }
}

export function canonicalize(value: unknown): string {
  const parts: string[] = [];
  emit(parts, value);
  return parts.join("");
}

export function canonicalBytes(value: unknown): Uint8Array {
  return new TextEncoder().encode(canonicalize(value));
}

function emit(out: string[], v: unknown): void {
  if (v === null) {
    out.push("null");
    return;
  }
  if (v === true) {
    out.push("true");
    return;
  }
  if (v === false) {
    out.push("false");
    return;
  }
  if (typeof v === "number") {
    if (!Number.isInteger(v)) {
      throw new CanonicalError("floating-point numbers are not allowed in signed records");
    }
    out.push(String(v));
    return;
  }
  if (typeof v === "bigint") {
    out.push(v.toString());
    return;
  }
  if (typeof v === "string") {
    emitString(out, v);
    return;
  }
  if (Array.isArray(v)) {
    out.push("[");
    for (let i = 0; i < v.length; i++) {
      if (i > 0) out.push(",");
      emit(out, v[i]);
    }
    out.push("]");
    return;
  }
  if (typeof v === "object") {
    const obj = v as Record<string, unknown>;
    const keys = Object.keys(obj).sort();
    out.push("{");
    for (let i = 0; i < keys.length; i++) {
      if (i > 0) out.push(",");
      emitString(out, keys[i]!);
      out.push(":");
      emit(out, obj[keys[i]!]);
    }
    out.push("}");
    return;
  }
  throw new CanonicalError(`unsupported value type: ${typeof v}`);
}

function emitString(out: string[], s: string): void {
  out.push('"');
  for (const ch of s) {
    const cp = ch.codePointAt(0)!;
    if (ch === '"') out.push('\\"');
    else if (ch === "\\") out.push("\\\\");
    else if (ch === "\b") out.push("\\b");
    else if (ch === "\f") out.push("\\f");
    else if (ch === "\n") out.push("\\n");
    else if (ch === "\r") out.push("\\r");
    else if (ch === "\t") out.push("\\t");
    else if (cp < 0x20) out.push("\\u" + cp.toString(16).padStart(4, "0"));
    else out.push(ch);
  }
  out.push('"');
}

File: sdk/py/cocore/canonical.py
"""Sorted-key canonical JSON serialization.

MUST produce byte-identical output to provider/src/canonical.rs and
packages/sdk/src/canonical.ts — the provider signs the canonical bytes with its
Secure Enclave key and this module re-canonicalizes to verify. Any divergence
breaks every signature.

Rules (matched across all three implementations): UTF-8, object keys sorted
lexicographically, no insignificant whitespace, minimal string escapes, integers
only (floats raise), booleans true/false, null allowed, bytes are base64 strings
before they reach this layer.
"""

from __future__ import annotations

from typing import Any


class CanonicalError(ValueError):
    pass


def canonicalize(value: Any) -> str:
    parts: list[str] = []
    _emit(parts, value)
    return "".join(parts)


def canonical_bytes(value: Any) -> bytes:
    return canonicalize(value).encode("utf-8")


def _emit(out: list[str], v: Any) -> None:
    if v is None:
        out.append("null")
    elif v is True:
        out.append("true")
    elif v is False:
        out.append("false")
    elif isinstance(v, int):  # note: bool handled above (bool is a subclass)
        out.append(str(v))
    elif isinstance(v, float):
        raise CanonicalError("floating-point numbers are not allowed in signed records")
    elif isinstance(v, str):
        _emit_string(out, v)
    elif isinstance(v, (list, tuple)):
        out.append("[")
        for i, item in enumerate(v):
            if i > 0:
                out.append(",")
            _emit(out, item)
        out.append("]")
    elif isinstance(v, dict):
        out.append("{")
        for i, key in enumerate(sorted(v.keys())):
            if i > 0:
                out.append(",")
            _emit_string(out, key)
            out.append(":")
            _emit(out, v[key])
        out.append("}")
    else:
        raise CanonicalError(f"unsupported value type: {type(v).__name__}")


_ESCAPES = {
    '"': '\\"',
    "\\": "\\\\",
    "\b": "\\b",
    "\f": "\\f",
    "\n": "\\n",
    "\r": "\\r",
    "\t": "\\t",
}


def _emit_string(out: list[str], s: str) -> None:
    out.append('"')
    for ch in s:
        esc = _ESCAPES.get(ch)
        if esc is not None:
            out.append(esc)
        elif ord(ch) < 0x20:
            out.append("\\u" + format(ord(ch), "04x"))
        else:
            out.append(ch)
    out.append('"')
